src/overlay/OverlayManagerImpl.cpp

```diff
@@ -346,32 +346,41 @@ OverlayManagerImpl::storePeerList(std::vector<PeerBareAddress> const& addresses,
                                   bool setPreferred, bool startup)
 {
     ZoneScoped;
-    auto typeUpgrade = setPreferred
-                           ? PeerManager::TypeUpdate::SET_PREFERRED
-                           : PeerManager::TypeUpdate::UPDATE_TO_OUTBOUND;
+    auto typeUpgrade = setPreferred ? PeerManager::TypeUpdate::SET_PREFERRED
```

Comment:
What the old interface was trying (and failed) to do was to have `PeerManager` be the one responsible for setting the right type in the database so that callers don't have to care about this.

I think what you did here (remove `UPDATE_TO_OUTBOUND`) but I don't think the other changes in this method are right:
`SET_OUTBOUND`  should really behave like `UPDATE_TO_OUTBOUND` as the mechanism to "downgrade" a preferred peer is to use `REMOVE_PREFERRED`.

That said the names are just wrong as they don't really convey what they're doing:
what the interface is really about is describing transitions based on claims not states.

So maybe this PR should really be about deleting `SET_OUTBOUND` and rename the other one to something like `IS_PREFERRED_OR_OUTBOUND` (as a preferred peer is "better than outbound").

While we're at it, we could also rename the other ones to `IS_PREFERRED` and `IS_NOT_PREFERRED`, and I think we'll have something that makes sense?

Reply:
I don't think what you propose works. `SET_OUTBOUND` behaving like `UPDATE_TO_OUTBOUND` was where I started. However, I do not think you can let PeerManager be solely responsible for figuring out the type, because we're using this API in two different places with different contracts. First one: update DB peers based on config, but do not downgrade any known preferred peers. Second: set a peer outbound if we tried connecting to it, regardless of its current type. https://github.com/stellar/stellar-core/blob/adfe86c5281432e6f653271b4e313e852b0099ff/src/overlay/Peer.cpp#L1031

From what I understand, the latter case has to be unconditional. Specifically, if the existing peer in the DB is preferred, but actually it is not, we have to downgrade it to outbound (I guess this can happen if we previously learned it's preferred, then re-started the node with updated preferred peers, removing that peer). So, basically, at the time of the function call you won't know what upgrade type you need until you load the peer from the database, and `SET_OUTBOUND` sort of handles it in a hacky way. Overall, this API is very confusing, and probably needs a much bigger cleanup, but I thought I'd start somewhere.

Reply:
Oh I missed that there was a nested conditional operator in `Peer::updatePeerRecordAfterEcho()` - this is evil :)

I still think that having more logic that deals with peer's state outside of "peer manager" is not a step forward as we're creating parallel logic to update state.
* If we wanted to update the state directly in code, we'd need to use the "peer" object (that represents a peer, and syncs with the database) - which would be a departure from the `PeerBareAddress` contracts.
    * Probably the right longer term move as it's more in line with how a data layer typically works, but a larger change compared to what we do now.
* Like you noted those transitions depend on the previous value, and this is exactly what `PeerManager::update` is supposed to do.
    * I think the more conservative change is to do what I was saying, which is "express facts" from the call site, and let "peer manager" to whatever it needs to do.

So, I think we should just change the interface of `PeerManager::update` to take better parameters:
* we can probably express all those intents with different methods instead of a generic "update"
* the path in "config" is really just trying to ensure that the entry is in the database
* the other path knows if the peer is preferred, so it should just be a parameter that peer manager can use to derive the actual values

Reply:
@MonsieurNicolas updated the diff. The updated API has users provide the desired peer type (so no direction needed, which I think makes it easier to reason about), and a flag for whether we definitely know that a peer is preferred.